app/javascript/mastodon/features/interaction_modal/index.jsx

```diff
@@ -18,62 +25,207 @@ const mapDispatchToProps = (dispatch) => ({
   },
 });
 
-class Copypaste extends React.PureComponent {
+const PERSISTENCE_KEY = 'mastodon_home';
+
+const isValidDomain = value => {
+  const url = new URL('https:///path');
+  url.hostname = value;
+  return url.hostname === value;
+};
+
+const valueToDomain = value => {
+  // If the user uses a URL to their profile page or server
+  if (/^https?:\/\//.test(value)) {
+    try {
+      return (new URL(value)).host;
+    } catch {
+      return '';
+    }
+  // If the user writes their full handle including username
+  } else if (value.includes('@')) {
+    const segments = value.split('@');
+
+    if (segments.length === 2) {
+      return segments[1];
+    } else {
+      return '';
```

Comment:
This'll cause problems if/when domain-only accounts are a thing; perhaps it'd be better to perform an account search here?

Reply:
There's another more immediate issue: this doesn't support situations where `LOCAL_DOMAIN` ≠ `WEB_DOMAIN`. Indeed, if `@foo@bar.name` is served by `https://social.bar.name/users/foo`, this whole flow redirects to `https://social.bar.name/authorize_interaction?uri=…`, which requires an extra redirect, while 3.x's interaction modal handled that case just fine.